src/unity/python/turicreate/toolkits/object_detector/object_detector.py

```diff
@@ -543,9 +545,8 @@ def numpy_worker():
             iteration = batch['iteration']
 
         # Wait for any pending batches and finalize our progress updates.
-        while batches_finished < batches_started:
-            batches_finished += 1
-            cur_loss = mps_net.wait_for_batch().sum() / mps_loss_mult
+        while len(batch_queue) > 0:
+            cur_loss = wait_for_batch()
```

Comment:
Shouldn't `update_progress(cur_loss, iteration)` be inside this while loop? The `update_progress` function computes a smoothed value of the loss, and the way this while loop is structured, it would seem that it could miss losses from being included in the running average.

Reply:
Not sure what would happen with `iteration` in that case though. The batch queue shouldn't really ever be more than 1 here though, right? If that's the case, then why the while loop?

Reply:
`update_progress` implicitly assumes that it's called only once per iteration, with the loss for the last batch in each iteration. In practice, `batch_queue` should have length 2 here almost all of the time, for the last two batches in the last iteration through the data (due to double buffering).

It looks like we'll do the wrong thing if the entire dataset fits into one batch, but I'm not going to worry about that now. Maybe I can clean up some of the mess I made here when we re-implement this logic in C++ for the C API version of OD.